src/hotspot/share/opto/vtransform.hpp

```diff
@@ -0,0 +1,516 @@
+class VTransform : public StackObj {
+private:
+  const VLoopAnalyzer& _vloop_analyzer;
+  const VLoop& _vloop;
+
+  NOT_PRODUCT(const VTransformTrace _trace;)
+
+  // Everything in the vtransform is allocated from this arena, including all vtnodes.
+  Arena _arena;
+
+  VTransformGraph _graph;
+
+  // Memory reference, and the alignment width (aw) for which we align the main-loop,
+  // by adjusting the pre-loop limit.
+  MemNode const* _mem_ref_for_main_loop_alignment;
+  int _aw_for_main_loop_alignment;
+
+public:
+  VTransform(const VLoopAnalyzer& vloop_analyzer,
+             MemNode const* mem_ref_for_main_loop_alignment,
+             int aw_for_main_loop_alignment
+             NOT_PRODUCT( COMMA const VTransformTrace trace)
+             ) :
+    _vloop_analyzer(vloop_analyzer),
+    _vloop(vloop_analyzer.vloop()),
+    NOT_PRODUCT(_trace(trace) COMMA)
+    _arena(mtCompiler),
+    _graph(_vloop_analyzer, _arena NOT_PRODUCT(COMMA _trace)),
+    _mem_ref_for_main_loop_alignment(mem_ref_for_main_loop_alignment),
+    _aw_for_main_loop_alignment(aw_for_main_loop_alignment) {}
+
+  const VLoopAnalyzer& vloop_analyzer() const { return _vloop_analyzer; }
+  Arena* arena() { return &_arena; }
+  bool is_empty() const { return _graph.is_empty(); }
+  VTransformGraph& graph() { return _graph; }
+
+  bool schedule() { return _graph.schedule(); }
+  void apply();
+
+private:
+  // VLoop accessors
+  PhaseIdealLoop* phase()     const { return _vloop.phase(); }
+  PhaseIterGVN& igvn()        const { return _vloop.phase()->igvn(); }
+  IdealLoopTree* lpt()        const { return _vloop.lpt(); }
+  CountedLoopNode* cl()       const { return _vloop.cl(); }
+  int iv_stride()             const { return cl()->stride_con(); }
+  bool in_bb(const Node* n)   const { return _vloop.in_bb(n); }
+
+  // VLoopVPointers accessors
+  const VPointer& vpointer(const MemNode* mem) const {
+    return _vloop_analyzer.vpointers().vpointer(mem);
+  }
+
+  // Ensure that the main loop vectors are aligned by adjusting the pre loop limit.
+  void determine_mem_ref_and_aw_for_main_loop_alignment();
+  void adjust_pre_loop_limit_to_align_main_loop_vectors();
+
+  void apply_vectorization() const;
+};
+
+// The vtnodes (VTransformNode) resemble the C2 IR Nodes, and model a part of the
+// VTransform. Many such vtnodes make up the VTransformGraph. The vtnodes represent
+// the resulting scalar and vector nodes as closely as possible.
+// See description at top of this file.
+class VTransformNode : public ArenaObj {
+public:
+  const VTransformNodeIDX _idx;
+
+private:
+  // _in is split into required inputs (_req), and additional dependencies.
+  const uint _req;
+  GrowableArray<VTransformNode*> _in;
+  GrowableArray<VTransformNode*> _out;
+
+public:
+  VTransformNode(VTransform& vtransform, const uint req) :
+    _idx(vtransform.graph().new_idx()),
+    _req(req),
+    _in(vtransform.arena(),  req, req, nullptr),
+    _out(vtransform.arena(), 4, 0, nullptr)
```

Comment:
Feel free to take this up with the Runtime team ;)

Reply:
There are advantages and disadvantages to having an explicit `filler`. Because the way `GrowableArray` is implemented there needs to be a `fillter`. I guess one could have a `default` filler - but then that would only work for some pre-defined types I think... complicated.

Reply:
For convenience, it could offer a constructor without filler which then just uses `nullptr`. But anyway, it's just a minor thing I've noticed while reviewing the code.